pitop/processing/algorithms/ball_detect.py

```diff
@@ -60,133 +92,72 @@ def __init__(self,
             atexit.register(self.__print_fps)
 
-    def detect(self, frame, color: Union[str, tuple] = "red"):
-        colors = self.__grab_colors(color)
-        cv_frame, resized_frame = self.__prepare_frames(frame)
-
-        ball_centers = {}
-        ball_radii = {}
-        ball_angles = {}
-        ball_finds = {}
-        robot_view = cv_frame.copy()
-        for color in colors:
-            ball_center, ball_radius = self.__find_most_likely_ball(color, resized_frame)
-
-            if ball_center is not None:
-                ball_center, ball_radius = self.__reposition_to_original_frame_size(ball_center, ball_radius)
+    def detect(self, frame, color: Union[str, list] = "red"):
+        def parse_colors(color_arg):
+            colors = []
+            if type(color_arg) == str:
+                assert(color_arg in valid_colors)
+                colors = [color_arg]
+            elif type(color_arg) in (list, tuple):
+                assert(set(color_arg).issubset(valid_colors))
+                colors = color_arg
+            if len(colors) > 3:
+                raise ValueError("Cannot pass more than three colors.")
 
-                self._detection_points[color].appendleft(ball_center)
+            return colors
 
-                self.__draw_ball_position(robot_view, color, ball_center, ball_radius)
-
-                ball_center = center_reposition(ball_center, robot_view)
-                ball_angle = get_object_target_lock_control_angle(ball_center, robot_view)
+        frame = ImageFunctions.convert(frame, format="OpenCV")
 
-                ball_centers[color] = ball_center
-                ball_radii[color] = ball_radius
-                ball_angles[color] = ball_angle
-                ball_finds[color] = True
-            else:
-                self._detection_points[color].appendleft(None)
-                ball_centers[color] = None
-                ball_radii[color] = None
-                ball_angles[color] = None
-                ball_finds[color] = False
+        if self._frame_scaler is None:
+            _, width = frame.shape[0:2]
+            self._frame_scaler = width / self._process_image_width
 
-            self.__draw_ball_contrail(robot_view, color)
+        for c in parse_colors(color):
+            self.balls[c] = self.__find_most_likely_ball(ball=self.balls.get(c),
+                                                         color=c,
+                                                         frame=frame)
 
-        if self.format.lower() != 'opencv':
-            robot_view = ImageFunctions.convert(robot_view, format="PIL")
+        robot_view = frame.copy()
+        ball_data = DotDict({})
+        for ball_color, ball_object in self.balls.items():
+            ball_data[ball_color] = ball_object
+            if ball_object.is_valid():
+                self.__draw_ball_position(robot_view, ball_object)
+                self.__draw_ball_contrail(robot_view, ball_object)
```

Comment:
This should be drawn regardless of if the ball object is valid, it will "fizzle" out anyway since "None" will be loaded into the deque of many ball detections are not found - this is the behaviour we want